code/extensions/SagePayExtension.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+use Omnipay\SagePay\Message\ServerNotifyResponse;
+use SilverStripe\Omnipay\Service\ServiceResponse;
+
+/**
+ * Sagepay has some very indiviual needs so to help we have created this
+ * extension that you can use as a beginning to try and help your SagePay
+ * transaction to go through. Add it to your YML file e.g.
+ *
+ * SilverStripe\Omnipay\Service\PaymentService:
+ *  extensions:
+ *    - SagePayExtension
+ *
+ * The extension adds following:
+ *  - A Description into the initial request (reequired by SagePau)
+ *  - The transactionReference which is recieved back from SagePaty
+ *  - A response to the notification (See use_async_notification)
+ * @see https://github.com/silverstripe/silverstripe-omnipay/issues/153
+ * @see https://github.com/silverstripe/silverstripe-omnipay/issues/159
+ */
+class SagePayExtension extends Extension
+{
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforePurchase(array &$gatewayData)
+    {
+        $this->addDescription($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeAuthorize(array &$gatewayData)
+    {
+        $this->addDescription($gatewayData);
+    }
+
+    /**
+     * @param ServiceResponse $response
+     */
+    public function updateServiceResponse($response)
+    {
+        $payment = $response->getPayment();
+
+        // We only want to respond to the notification if we are using SagePay
+        if ($payment->Gateway !== 'SagePay_Server') {
+            return;
+        }
+
+        $this->respondToNotification($response, $payment);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeCompletePurchase(array &$gatewayData)
+    {
+        $gatewayData['transactionReference'] = $this->getTransactionReference($gatewayData);
+    }
+
+    /**
+     * Grabs the transactionReference from the previous received message so
+     * that it can be sent back to SagePay as clarification
+     * @return string
+     */
+    private function getTransactionReference()
+    {
+        /** @var \Payment $payment */
+        $payment = $$this->owner->purchaseService->getPayment();
```

Comment:
Is this a typo? `$$this` instead of `$this`?
I think `$this->owner->getPayment()` should be enough?